FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+            matchFound = matchEndWildcards( pTopicFilter,
+                                            topicNameLength,
+                                            topicFilterLength,
+                                            nameIndex,
+                                            filterIndex,
+                                            &status );
+        }
+        else
+        {
+            /* Check for matching wildcards. */
+            matchFound = matchWildcards( pTopicFilter,
+                                         pTopicName,
+                                         topicNameLength,
+                                         filterIndex,
+                                         &nameIndex,
+                                         &status );
+        }
+
+        if( matchFound == true )
+        {
+            break;
+        }
+
+        /* Increment indexes. */
+        nameIndex++;
+        filterIndex++;
+    }
+
+    if( status == false )
+    {
+        /* If the end of both strings has been reached, they match. */
+        status = ( ( nameIndex == topicNameLength ) && ( filterIndex == topicFilterLength ) );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static void prvSubscriptionManager( MQTTContext_t * pMqttContext,
+                                    MQTTPacketInfo_t * pPacketInfo,
+                                    uint16_t packetIdentifier,
+                                    MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pMqttContext != NULL );
+    assert( pPacketInfo != NULL );
+    AckInfo_t xAckInfo;
+    MQTTStatus_t xStatus = MQTTSuccess;
+    bool xIsMatched = false;
+    size_t i;
+    MQTTSubscribeInfo_t * pxSubscribeInfo = NULL;
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
+    {
+        assert( pPublishInfo != NULL );
+
+        for( i = 0; i < SUBSCRIPTIONS_MAX_COUNT; i++ )
+        {
+            if( pxSubscriptions[ i ].topicFilterLength > 0 )
+            {
+                xIsMatched = topicFilterMatch( pPublishInfo->pTopicName,
+                                               pPublishInfo->topicNameLength,
+                                               pxSubscriptions[ i ].pTopicFilter,
+                                               pxSubscriptions[ i ].topicFilterLength );
+
+                if( xIsMatched )
+                {
+                    LogInfo( ( "Adding publish to response queue for %.*s",
+                               pxSubscriptions[ i ].topicFilterLength,
+                               pxSubscriptions[ i ].pTopicFilter ) );
+                    prvCopyPublishToQueue( pPublishInfo, pxSubscriptions[ i ].pResponseQueue );
+                }
+            }
+        }
+    }
+    else
```

Comment:
I think everything in else should be handled in a different function, as the name prvSubscriptionManager hints that it is only dealing the subscription management. How about calling prvSubscriptionManager only for an incoming publish? All other incoming packets can be handled in a different function.

Reply:
I thought of that, but it does manage subscriptions due to incoming subacks/unsubacks, and the only packets not dealing with subscriptions is the publish acks. What would you suggest would be a better name for this function?

Reply:
I see. I still think separating out the subscription manager logic and the other acks will be a good addition instead of renaming this function. How about only handling the incoming publishes, SUBACK and UNSUBACK in the subscription manager logic?